reactjs/reactjs.org
# [beta] Sandpack - new bundler

## Sandpack bundler v2 - demo

The Sandpack bundler v2 aims to eventually replace the current sandpack with a more streamlined and faster version. Plus, it allows us to implement even more improvements in the future, thanks to a fresh codebase that allowed us to remove many of the bottlenecks from the previous bundler. 

**The new bundler:** 

- Skips transpiling step for dependencies by caching them in our custom-built CDN;
- Fetches dependencies from our very own CDN, no longer third-party involved;
- Uses less resource (avoid fetching unnecessary files) and lot quicker;

<details>
<summary><strong>Same benefits from the previous bundler</strong></summary>


**Security**

The bundler evaluates and transpiles all files in an iframe under a different subdomain. This is important because it prevents attackers from tampering with cookies of the host domain when evaluating code.

**Scoped styles/javascript execution**

There’s no risk of any executed code affecting the main page, which avoids unexpected styles or unhandled errors.

**Performance**

We heavily make use of Web Workers for transpilations, and only for those sandboxes that are in the viewport and the user is interacting with, meaning that we only use resources as needed.

**Bundle Size**

Another reason to host the bundler externally is because of code splitting: we split all our transpilers away and load them on-demand.

**React DevTools**

Sandpack bundler supports running `react-devtools-inline` in isolated mode, as an external dependency, which means you can run more than one instance per page and lazily load it.

**Others**

- npm dependency support;
- Hot module reloading;
- Error overlaying;
- Caching;
</details>

<details>
<summary><strong>Benchmarketing</strong></summary>

| | v1 | v2 |
| - | - | - |
| Memory usage (iframe thread) | ~20MB | ~ 5MB |
| Memory usage (babel) [1] | 6.8 - 10.2MB | 6.6 - 6.6MB |
| Network (compressed) [2] | 2.8MB | 863kb |
| Time to load (first time) [3] | 9293ms | 4149ms |
| Time to load (second time) [4] | 565ms | 566ms |
| Time to load [5] | 854ms | 214ms |

[1] idle - peek
[2] Exclude local resource & external, except sandpack & disabled cache;
[3] Incognito window (fresh cache) & Fast 3G 
[4] Cache enabled & Fast 3G 
[5] Incognito window (fresh cache) & no throttling

</details>

<details>
<summary><strong>Known issues / what’s next</strong></summary>

- [x] <s>Implement `react-devtools-inline`  protocol</s>;
- [x] <s>Server Components (still on discovery phase)</s>;

</details>


<details>
<summary><strong>How to test it & how to use it</strong></summary>

**How to test**
This PR is completely disposable, and it's only for testing purposes. So, in order to test it follow the steps:
- Deploy this PR;
- Access `/sandpack` page or this example page https://reactjs-org-do4znpw6x-danilowoz.vercel.app/sandpack;
- Tick the checkbox to switch between bundle versions **for all pages**;
- Feel free to navigate between pages and note that there's a timer to count how long a sandbox takes to load;

**How to use**

```diff
<SandpackProvider
	...
+   bundlerURL="https://cf56a978.sandpack-bundler.pages.dev"
 >
```

</details>

---

Hope it addresses most of the problems we've discussed before. We're looking forward to hearing any feedback about this new version and please let us know about any problem you notice.

Kudos to @DeMoorJasper for all the hard work on this! 🙌
